### ISSUE TITLE
react.gradle does not support custom release build types (name not including "release")

### ISSUE THREAD

**Issue by draperunner** [facebook/react-native#30788]
## Description
We have several Android release build types, defined in our `android/app/build.gradle`. They have other names than Release, of course. Like "Staging".

The problem with `react.gradle` is that it has multiple checks for `isRelease` that only consider the _target name_, like here:
https://github.com/facebook/react-native/blob/164e133a98fade1c0f7577476e411bf9f6c2b7ca/react.gradle#L330

As a consequence of this, Hermes sourcemaps are not built for these release variants, as `hermesFlagsDebug` is used, not `hermesFlagsRelease`: https://github.com/facebook/react-native/blob/164e133a98fade1c0f7577476e411bf9f6c2b7ca/react.gradle#L189

Maybe one could instead use `"config.bundleIn${targetName}"` and/or `config."devDisabledIn${targetName}"` to check if a target is a release variant?

## React Native version:
`0.63.4`

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Add another buildType block to `android/app/build.gradle` called "Staging". Something like this:
```staging {
            initWith(buildTypes.release)
            applicationIdSuffix ".staging"
            matchingFallbacks = ['release']
        }
```
2. Configure Staging for release in `project.ext.react`:
```
project.ext.react = [
    devDisabledInRelease : true,
    bundleInRelease      : true,
    devDisabledInStaging : true,
    bundleInStaging      : true,
```
3. Add some logging to `node_modules/react-native/react.gradle` to check the evaluation of `isRelease`
```diff
def isRelease = targetName.toLowerCase().contains("release")
+println "isRelease? ${isRelease}"
```
4. Build the app with gradle and check the logs
```
cd android
./gradlew assembleStaging
```

## Expected Results
I expect other release variants to also be treated as the one called Release.

## Snack, code example, screenshot, or link to a repository:

Simplified and redacted. We use Yarn Workspaces (monorepo) so paths might look weird.

`android/app/build.gradle`:
```
project.ext.react = [
    root                 : "../..",
    cliPath              : "../../node_modules/react-native/cli.js",
    bundleConfig         : "metro.config.js",
    composeSourceMapsPath: "../../node_modules/react-native/scripts/compose-source-maps.js",
    hermesCommand        : "../../../../node_modules/hermes-engine/%OS-BIN%/hermesc",
    entryFile            : "index.ts",
    enableHermes: true,  // clean and rebuild if changing

    environmentForRelease: "prod",
    devDisabledInRelease : true,
    bundleInRelease      : true,
    environmentForStaging: "staging",
    devDisabledInStaging : true,
    bundleInStaging      : true,
]
```
```
buildTypes {
        release {
            minifyEnabled enableProguardInReleaseBuilds
        }
        staging {
            initWith(buildTypes.release)
            applicationIdSuffix ".staging"
            matchingFallbacks = ['release']
        }
```

**Comment by draperunner** [comment]
If this is a valid approach, I would happily submit a PR

```diff
-def isRelease = (targetName.toLowerCase().contains("release")
+def isRelease = (targetName.toLowerCase().contains("release") || config."devDisabledIn${targetName}" && config."bundleIn${targetName}")
```

**Comment by yuki-js** [comment]
I have same issue.

**Comment by kabbi** [comment]
I have the same issue. There is also another line in build.gradle, that disables source map and optimizations in non-release build targets
https://github.com/facebook/react-native/blob/ace025d2cbcc06ef076cbf460be445b18baddcc7/react.gradle#L180

**Comment by jittuu** [comment]
@draperunner did you find any workaround? I have the same issue with sentry.

**Comment by draperunner** [comment]
@jittuu We're using the change I suggested in my previous comment. We are patching `react.gradle` on postinstall using the technique explained here: https://opensource.christmas/2019/4

**Comment by yuki-js** [comment]
I solved it by making buildType `Stgrelease`, which contains `release` and worked as expected.

**Comment by cortinico** [comment]
Closing as this has been fully addressed in 0.71
